### PR TITLE
48 feature define action event record schema and json serialization for cli output

### LINKED ISSUE #48: [Feature] Define action event record schema and JSON serialization for CLI output
### Goal

Define the JSON output schema for Sprint 2 and implement serialization of detection results into a stable log format.

### Scope

In scope:
- define JSON record fields such as frame range or timestamp, label, confidence, optional `track_id`
- implement JSON writer
- validate output structure
- document example output

### Definition of Done

- [x] JSON schema is documented
- [x] JSON writer is implemented
- [x] Sample output file is generated locally
- [x] Invalid or empty outputs are handled cleanly
- [x] Tested locally
- [x] Docs/README updated if relevant
- [x] PR created and linked to this issue

### PR DESCRIPTION
**Pull request by Szymon110903**
## Summary
Describe what was changed and why.

## Linked Issue
Closes #48 

## Checklist
- [x] Changes were tested locally
- [x] CI passes
- [x] README/docs updated if relevant
- [x] Scope matches the linked Issue
- [x] No direct work outside the agreed scope

### CODE REVIEW

File: tests/inference/test_modules.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+import sys
+import json
+from pathlib import Path
+
+# Add root directory to path (go up 3 levels: tests/inference -> tests -> . -> .)
+sys.path.insert(0, str(Path(__file__).parent.parent.parent))
+
+from src.inference.action_event import ActionEvent, ActionEventLog
+from src.inference.json_writer import ActionEventWriter
```
[the remark this answers is not in the document]
Comment (Szymon110903) [reply tests/inference/test_modules.py]: @copilot apply changes based on [this feedback](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/pull/59#discussion_r3059894754)

File: src/inference/json_writer.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        """
+        if track_ids is None:
+            track_ids = [None] * len(results)
+
+        if len(track_ids) != len(results):
+            logger.warning(
+                f"track_ids length ({len(track_ids)}) != results length ({len(results)})"
+            )
+            track_ids = [None] * len(results)
+
```
[the remark this answers is not in the document]
Comment (Szymon110903) [reply src/inference/json_writer.py:125]: @copilot apply changes based on [this feedback](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/pull/59#discussion_r3059894783)

File: src/inference/json_writer.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                label = str(prediction["label"])
+                confidence = float(prediction["confidence"])
+                return label, confidence
+
+        # Handle single confidence value (assume positive class)
+        if isinstance(prediction, (int, float)):
+            confidence = float(prediction)
+            label = "action" if confidence > 0.5 else "no_action"
+            return label, abs(confidence)
+
```
[the remark this answers is not in the document]
Comment (Szymon110903) [reply src/inference/json_writer.py:180]: @copilot apply changes based on [this feedback](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/pull/59#discussion_r3059894802)

File: src/inference/json_writer.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                probs = self._softmax(pred_array)
+
+            class_idx = int(np.argmax(probs))
+            confidence = float(np.max(probs))
+            label = self._get_label(class_idx)
+
+            return label, confidence
+
+        except Exception as e:
+            raise ValueError(f"Failed to extract from logits/probs: {e}")
```
[the remark this answers is not in the document]
Comment (Szymon110903) [reply src/inference/json_writer.py:244]: @copilot apply changes based on [this feedback](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/pull/59#discussion_r3059894819)

File: docs/backend.md [hunk 5]
```diff
@@ -3,6 +3,102 @@
...
+| Field | Type | Required | Description |
+|-------|------|----------|-------------|
+| `start_frame_index` | integer | Yes | Starting frame index of the detection window (0-indexed) |
+| `end_frame_index` | integer | Yes | Ending frame index of the detection window (inclusive) |
+| `label` | string | Yes | Label of the detected action/behavior class (e.g., "walking", "running") |
+| `confidence` | float | Yes | Confidence score of the prediction (0.0 to 1.0) |
+| `start_timestamp` | float | No | Starting timestamp in seconds |
+| `end_timestamp` | float | No | Ending timestamp in seconds |
+| `track_id` | integer | No | Tracking ID for multi-object tracking |
+
```
[the remark this answers is not in the document]
Comment (Szymon110903) [reply docs/backend.md:37]: @copilot apply changes based on [this feedback](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/pull/59#discussion_r3059894835)

**Comment by Szymon110903** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/lukaszjecek/human-behavior-recognition-in-video-streams/pull/59#pullrequestreview-4084582114)

**Review by Szymon110903 — COMMENTED** [review]
small fixes made by copilot

**Comment by lukaszjecek** [comment]
Scope issue #48 is mostly covered, but there are still a few things to fix before merge.

1. In `src/inference/json_writer.py`, the handling of numerical predictions is still inconsistent: the label is selected based on the raw value (`confidence > 0.5`), but the returned confidence is `abs(confidence)`. This can produce results such as `no_action` with confidence `0.9`, and values `> 1.0` are only rejected later by `ActionEvent` instead of being handled cleanly at the writer level.
2. In `src/inference/json_writer.py`, the type of `track_ids` in `add_results()` still does not match the implementation. The signature states `Optional[list[int]]`, but the code also operates on `None` values inside the list.
3. `docs/backend.md` still points to the old validation script path: `tests/inference/test_modules.py`, while the file was moved to `scripts/validate_modules.py`.
4. There is also an inconsistency around the sample output location: the validation script generates output under `scripts/data/sample_actions.json`, but the committed sample file in the repo is `tests/inference/data/logs/sample_actions.json`. Please fix this so it is clear which file is the canonical sample output and so that the correct location is clearly visible in the repository structure and documentation.
5. The PR body still lacks a formal issue link (`Closes #48`), which is part of the DoD for the issue.
6. The checkboxes have not been updated, either in the issue or in the PR, so it is still unclear from the tracking itself what has actually been delivered.

Once these points are addressed, I will be able to approve it.

**Comment by blanqtoja** [comment]
- [x] In `src/inference/action_event.py`, inside `__post_init__` (line 43), you check `isinstance` for the `label`, but ignore the rest of the parameters. An invalid case could occur if someone sends `{"confidence": "0.95"}` from the JSON API – a string will be assigned instead of a float.

- [x] There is no timestamp validation in `src/inference/action_event.py` `__post_init__`. What if someone provides a `start_timestamp` < 0? Consider using the Pydantic library, which provides built-in validators.

**ActionEventLog**
- [x] Lack of validation in `add_event`. An invalid case could happen if someone uses it this way: `log.add_event({"label": "zaq1"})`. Please add an `isinstance` check. Type hints in Python do not work at the interpreter level during runtime; they are only for warning developers in the IDE.

- [x] In the `to_dict` function, I can see that you export `event_count` (probably for future validation). However, in `load_from_file`, you ignore this field. You should either add an expression that raises an error when the loaded log count doesn't match this field, or just delete the `event_count` field from the `to_dict` function.

**Review by lukaszjecek — CHANGES_REQUESTED** [review]
Scope issue #48 is close, but there are still a few things to fix before merge.

### 1. The PR body still lacks a formal issue link (`Closes #48`), which is part of the DoD for this issue.
Image

### 2. There is still an inconsistency around the sample output location. The validation script generates the sample file under `scripts/data/sample_actions.json`, while the committed sample output in the repository is stored in `tests/inference/data/logs/sample_actions.json`. **Please make one location canonical and align the script, repository structure, and documentation with it.**
### 3. Please also update the checklist both in the issue and in the PR so it clearly shows what has already been completed.
Image

### 4. This branch is currently out of date with main, so please update it from the base branch before merge. 
Image

Once these points are addressed, I will be able to approve it.

**Review by lukaszjecek — APPROVED** [review]

**Review by blanqtoja — APPROVED** [review]
